proposals/0005-fallback-rounting.md

```diff
@@ -0,0 +1,171 @@
+---
+GEP: 5
+Title: Fallback & Routing
+Discussion: Link
+Implementation: Link
+---
+
+# Fallback & Routing
+
+## Abstract
+
+In this GEP, we talk about fallback mechanism, the key strategy to cope with model provider availability issues, 
+and load balancing which is a somewhat related topic.
+
+## Motivation
+
+Model provider APIs are external resources that **could be unavailable** from time to time.
+When your applications use those API in the business-critical workflows, this becomes a problem that could cause revenue loss.
+
+Model API could be unavailable in the following ways:
+- API returns `50X` errors that indicate internal service issues
+- API returns `429` errors indicating rate or quota limits. API is healthy in this case, but given credentials are not going to work for specified period of time
+- API could answer for too long a.k.a. **timeout**. Increase in timeout rate means availability issues on the provider side.
+
+At the same time, routing requests to a healthy model provider could be just one of requirements
+the application may have to pick the next model to serve the request. Other considerations may be helpful, too, like:
+- **Weighted Routing (a.k.a Weighted Round Robin or WRR)** e.g. let 30% of requests be processed by the Cohere model, 20% by Anthropic, and the remaining 50% by OpenAI.
+- **Priority-based Routing** e.g. OpenAI should serve all requests if available. If not, then Anthropic, and finally, Google Bard
+- **Latency-based Routing** e.g. the next request should be served by the fastest available model in the pool. This is useful for real-time applications where latency is important.
+- **Cost-based Routing** e.g. the next request should be served by the cheapest available model in the pool
```

Comment:
A feature for post-MVP might be some sort of budgeting mechanism. It would combine the cost and weighted routing mechanisms. For example, end user doesn't want to spend more than $10/day on openai once that threshold is hit pass the call to another provider. Or instead of implementing a threshold, we could balance call throughout the day (or whatever time period) so that a given cost threshold isn't exceeded.

Again, not necessary for MVP.

Reply:
Like using budgets as weights to determine the portion of traffic each model should serve? Yeah, I think this is a cool idea. As soon as start tracking budgets this should be easily implemented.